### PR TITLE
Removed UpdateOnCreate.  Refactored AuditLogger, Pagination, & Sorting

### PR DESCRIPTION
**Pull request by aklish**
A few fixes:
 - Fixes Issue #751
 - Added FilterExpression to DataStoreTransaction.catPaginate interface
 - Removed UpdateOnCreate check
 - Refactored Sorting
 - Refactored Pagination

## License
I confirm that this contribution is made under an Apache 2.0 license and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: elide-core/src/main/java/com/yahoo/elide/audit/AuditLogger.java [hunk 1]
```diff
@@ -27,7 +25,7 @@ public void log(LogMessage message) {
         messages.get().add(message);
     }
 
-    public abstract void commit(RequestScope requestScope) throws IOException;
+    public abstract void commit() throws IOException;
```
Comment (wcekan) [inline elide-core/src/main/java/com/yahoo/elide/audit/AuditLogger.java]: @aklish Currently, our audit logger uses requestScope to access the transaction in order to save audit record information to the database.  Is there another way then to access request scope?

File: elide-core/src/main/java/com/yahoo/elide/audit/LogMessageImpl.java [hunk 2]
```diff
@@ -0,0 +1,175 @@
+/*
+ * Copyright 2015, Yahoo Inc.
```
Comment (jkusa) [inline elide-core/src/main/java/com/yahoo/elide/audit/LogMessageImpl.java]:
```suggestion
 * Copyright 2020, Yahoo Inc.
```

File: elide-core/src/main/java/com/yahoo/elide/audit/LogMessage.java [hunk 3]
```diff
@@ -1,202 +1,54 @@
...
-     * @throws InvalidSyntaxException if the Audit annotation has invalid syntax.
+     * Gets operation code.
+     *
+     * @return the operation code
      */
-    public LogMessage(Audit audit, PersistentResource record, Optional<ChangeSpec> changeSpec)
-            throws InvalidSyntaxException {
-        this(audit.logStatement(), record, audit.logExpressions(), audit.operation(), changeSpec);
-    }
+    public int getOperationCode();
```
Comment (jkusa) [inline elide-core/src/main/java/com/yahoo/elide/audit/LogMessage.java]: What is an operation code in this context?

File: elide-core/src/test/java/com/yahoo/elide/audit/LogMessageTest.java [hunk 4]
```diff
@@ -73,15 +73,15 @@ public String getName() {
     @Test
     public void verifyOpaqueUserExpressions() {
         final String[] expressions = { "${opaqueUser.name}", "${opaqueUser.name}" };
-        final LogMessage message = new LogMessage("{0} {1}", childRecord, expressions, 1, Optional.empty());
+        final LogMessageImpl message = new LogMessageImpl("{0} {1}", childRecord, expressions, 1, Optional.empty());
```
Comment (jkusa) [inline elide-core/src/test/java/com/yahoo/elide/audit/LogMessageTest.java]: Can we rename the file to `LogMessageImplTest.java`?

File: elide-core/src/main/java/com/yahoo/elide/core/PersistentResource.java [hunk 5]
```diff
@@ -1768,7 +1768,7 @@ protected void auditField(final ChangeSpec changeSpec) {
         }
         for (Audit annotation : annotations) {
             if (annotation.action().length == 1 && annotation.action()[0] == Audit.Action.UPDATE) {
-                LogMessage message = new LogMessage(annotation, this, Optional.of(changeSpec));
+                LogMessageImpl message = new LogMessageImpl(annotation, this, Optional.of(changeSpec));
```
Comment (jkusa) [inline elide-core/src/main/java/com/yahoo/elide/core/PersistentResource.java]:
```suggestion
                LogMessage message = new LogMessageImpl(annotation, this, Optional.of(changeSpec));
```

Comment (jkusa) [reply elide-core/src/main/java/com/yahoo/elide/core/PersistentResource.java]: Somewhat out of scope, could this object creation be injected or provided by a factory class?

File: elide-core/src/main/java/com/yahoo/elide/core/PersistentResource.java [hunk 6]
```diff
@@ -1791,7 +1791,7 @@ protected void auditClass(Audit.Action action, ChangeSpec changeSpec) {
         for (Audit annotation : annotations) {
             for (Audit.Action auditAction : annotation.action()) {
                 if (auditAction == action) { // compare object reference
-                    LogMessage message = new LogMessage(annotation, this, Optional.ofNullable(changeSpec));
+                    LogMessageImpl message = new LogMessageImpl(annotation, this, Optional.ofNullable(changeSpec));
```
Comment (jkusa) [inline elide-core/src/main/java/com/yahoo/elide/core/PersistentResource.java]:
```suggestion
                    LogMessage message = new LogMessageImpl(annotation, this, Optional.ofNullable(changeSpec));
```

File: elide-core/src/main/java/com/yahoo/elide/jsonapi/EntityProjectionMaker.java [hunk 7]
```diff
@@ -136,11 +138,16 @@ public EntityProjection parseInclude(Class<?> entityClass) {
             Class<?> entityClass = getEntityClass(parentClass, entityName);
             FilterExpression filter = scope.getExpressionForRelation(parentClass, entityName).orElse(null);
 
+            Sorting sorting = SortingImpl.getDefaultEmptyInstance();
+            if (scope.getQueryParams().isPresent()) {
+                sorting = SortingImpl.parseQueryParams(scope.getQueryParams().get(), entityClass, dictionary);
```
Comment (jkusa) [inline elide-core/src/main/java/com/yahoo/elide/jsonapi/EntityProjectionMaker.java]: Could this be simplified with:
```java
Sorting sorting = SortingImpl.parseQueryParams(scope.getQueryParams().get(), entityClass, dictionary);
```

File: elide-core/src/main/java/com/yahoo/elide/core/sort/SortingImpl.java [hunk 8]
```diff
@@ -25,25 +27,29 @@
...
-
-    /**
-     * Denotes the intended sort type from json-api field.
-     */
-    public enum SortOrder { asc, desc }
+public class SortingImpl implements Sorting {
 
     private final Map<String, SortOrder> sortRules = new LinkedHashMap<>();
-    private static final Sorting DEFAULT_EMPTY_INSTANCE = null;
+    private static final SortingImpl DEFAULT_EMPTY_INSTANCE = null;
```
Comment (jkusa) [inline elide-core/src/main/java/com/yahoo/elide/core/sort/SortingImpl.java]: Do we need this? Can simply return null in `parseSortRules`?

File: elide-core/src/main/java/com/yahoo/elide/jsonapi/EntityProjectionMaker.java [hunk 9]
```diff
@@ -261,11 +268,16 @@ public EntityProjection visitIncludePath(Path path) {
                 filter = scope.getExpressionForRelation(parentClass, collectionNameText).orElse(null);
             }
 
+            Sorting sorting = SortingImpl.getDefaultEmptyInstance();
```
Comment (jkusa) [inline elide-core/src/main/java/com/yahoo/elide/jsonapi/EntityProjectionMaker.java]: Same here?

File: elide-core/src/main/java/com/yahoo/elide/request/Sorting.java [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+     * Get the type of the collection to sort.
+     * @return the collection type.
+     */
+    public Class<?> getType();
+
+    /**
+     * Is this sorting the default instance (not present).
+     * @return true if sorting wasn't requested.  False otherwise.
+     */
+    public boolean isDefaultInstance();
```
Comment (jkusa) [inline elide-core/src/main/java/com/yahoo/elide/request/Sorting.java]: Would this be better named `isEmpty()`

File: elide-core/src/main/java/com/yahoo/elide/ElideSettingsBuilder.java [hunk 11]
```diff
@@ -45,8 +45,8 @@
     private List<JoinFilterDialect> joinFilterDialects;
     private List<SubqueryFilterDialect> subqueryFilterDialects;
     private Map<Class, Serde> serdes;
-    private int defaultMaxPageSize = Pagination.MAX_PAGE_LIMIT;
-    private int defaultPageSize = Pagination.DEFAULT_PAGE_LIMIT;
+    private int defaultMaxPageSize = PaginationImpl.MAX_PAGE_LIMIT;
+    private int defaultPageSize = PaginationImpl.DEFAULT_PAGE_LIMIT;
```
Comment (jkusa) [inline elide-core/src/main/java/com/yahoo/elide/ElideSettingsBuilder.java]: Should these constants be defined in the interface?

File: elide-core/src/main/java/com/yahoo/elide/core/pagination/PaginationImpl.java [hunk 12]
```diff
@@ -121,7 +128,7 @@ public static Pagination fromOffsetAndLimit(int limit, int offset, boolean gener
             pageData.put(PAGE_KEYS.get(PAGE_TOTALS_KEY), 1);
         }
 
-        Pagination result = new Pagination(pageData.build(), MAX_PAGE_LIMIT, DEFAULT_PAGE_LIMIT);
+        PaginationImpl result = new PaginationImpl(pageData.build(), MAX_PAGE_LIMIT, DEFAULT_PAGE_LIMIT);
```
Comment (jkusa) [inline elide-core/src/main/java/com/yahoo/elide/core/pagination/PaginationImpl.java]:
```suggestion
        Pagination result = new PaginationImpl(pageData.build(), MAX_PAGE_LIMIT, DEFAULT_PAGE_LIMIT);
```

File: elide-core/src/main/java/com/yahoo/elide/request/Pagination.java [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+     * Set the total size of the collection.
+     * @param pageTotals the total size.
+     */
+    void setPageTotals(long pageTotals);
+
+    /**
+     * Is this the default instance (not present).
+     * @return true if pagination wasn't requested.  False otherwise.
+     */
+    public boolean isDefaultInstance();
```
Comment (jkusa) [inline elide-core/src/main/java/com/yahoo/elide/request/Pagination.java]: `isEmpty?` `hasNone?`

File: elide-core/src/test/java/com/yahoo/elide/core/PaginationLogicTest.java [hunk 14]
```diff
@@ -14,7 +14,7 @@
 import com.yahoo.elide.ElideSettingsBuilder;
 import com.yahoo.elide.annotation.Paginate;
 import com.yahoo.elide.core.exceptions.InvalidValueException;
-import com.yahoo.elide.core.pagination.Pagination;
+import com.yahoo.elide.core.pagination.PaginationImpl;
```
Comment (jkusa) [inline elide-core/src/test/java/com/yahoo/elide/core/PaginationLogicTest.java]: Should we rename the file to `PaginationImplTest.java`

File: elide-core/src/main/java/com/yahoo/elide/ElideSettingsBuilder.java (hunk 11 above)
Comment (aklish) [reply elide-core/src/main/java/com/yahoo/elide/ElideSettingsBuilder.java]: 👍

File: elide-core/src/main/java/com/yahoo/elide/audit/LogMessage.java (hunk 3 above)
Comment (aklish) [reply elide-core/src/main/java/com/yahoo/elide/audit/LogMessage.java]: The original audit design allowed for all log messages to have an audit code.  Most audit systems have something like this (in addition to a log message).

File: elide-core/src/main/java/com/yahoo/elide/core/PersistentResource.java (hunk 5 above)
Comment (aklish) [reply elide-core/src/main/java/com/yahoo/elide/core/PersistentResource.java]: 👍

File: elide-core/src/main/java/com/yahoo/elide/core/sort/SortingImpl.java (hunk 8 above)
Comment (aklish) [reply elide-core/src/main/java/com/yahoo/elide/core/sort/SortingImpl.java]: This makes sorting consistent with pagination and allows us to introduce a default Sort annotation similar to what we have for Pagination in the future.

File: elide-core/src/main/java/com/yahoo/elide/jsonapi/EntityProjectionMaker.java (hunk 7 above)
Comment (aklish) [reply elide-core/src/main/java/com/yahoo/elide/jsonapi/EntityProjectionMaker.java]: That doesn't cover the case where there are no query params.   I believe the code is as simple as can be.

File: elide-core/src/main/java/com/yahoo/elide/jsonapi/EntityProjectionMaker.java (hunk 9 above)
Comment (aklish) [reply elide-core/src/main/java/com/yahoo/elide/jsonapi/EntityProjectionMaker.java]: See above.

File: elide-core/src/main/java/com/yahoo/elide/request/Sorting.java (hunk 10 above)
Comment (aklish) [reply elide-core/src/main/java/com/yahoo/elide/request/Sorting.java]: Default may not equal empty in the future.

File: elide-core/src/test/java/com/yahoo/elide/audit/LogMessageTest.java (hunk 4 above)
Comment (aklish) [reply elide-core/src/test/java/com/yahoo/elide/audit/LogMessageTest.java]: :+1:

File: elide-core/src/test/java/com/yahoo/elide/core/PaginationLogicTest.java (hunk 14 above)
Comment (aklish) [reply elide-core/src/test/java/com/yahoo/elide/core/PaginationLogicTest.java]: 👍

File: elide-core/src/main/java/com/yahoo/elide/audit/LogMessage.java (hunk 3 above)
Comment (jkusa) [reply elide-core/src/main/java/com/yahoo/elide/audit/LogMessage.java]: This is a generic value correct? Can we mention that in the javadocs?

File: elide-core/src/main/java/com/yahoo/elide/jsonapi/EntityProjectionMaker.java (hunk 7 above)
Comment (jkusa) [reply elide-core/src/main/java/com/yahoo/elide/jsonapi/EntityProjectionMaker.java]: Can this cover that case:
https://github.com/yahoo/elide/pull/1146/files#diff-9f7716d84a63e43ad20f499737c723cdR152

File: elide-core/src/main/java/com/yahoo/elide/audit/LogMessageImpl.java (hunk 2 above)
Comment (aklish) [reply elide-core/src/main/java/com/yahoo/elide/audit/LogMessageImpl.java]: changed

File: elide-core/src/main/java/com/yahoo/elide/audit/LogMessage.java (hunk 3 above)
Comment (aklish) [reply elide-core/src/main/java/com/yahoo/elide/audit/LogMessage.java]: 👍

File: elide-core/src/main/java/com/yahoo/elide/core/PersistentResource.java (hunk 6 above)
Comment (aklish) [reply elide-core/src/main/java/com/yahoo/elide/core/PersistentResource.java]: Already changed

File: elide-core/src/main/java/com/yahoo/elide/jsonapi/EntityProjectionMaker.java (hunk 7 above)
Comment (aklish) [reply elide-core/src/main/java/com/yahoo/elide/jsonapi/EntityProjectionMaker.java]: Check out the new refactor.

**Pull request merged**